haystack/nodes/ranker/most_diverse.py

```diff
@@ -0,0 +1,146 @@
+        devices: Optional[List[Union[str, "torch.device"]]] = None,
+    ):
+        """
+        Initialize a MostDiverseRanker.
+
+        :param model_name_or_path: Path to a pretrained sentence-transformers model.
+        :param use_gpu: Whether to use GPU (if available). If no GPUs are available, it falls back on a CPU.
+        :param devices: List of torch devices (for example, cuda:0, cpu, mps) to limit inference to specific devices.
+        """
+        torch_and_transformers_import.check()
+        super().__init__()
+        self.devices, _ = initialize_device_settings(devices=devices, use_cuda=use_gpu, multi_gpu=True)
+        self.model = SentenceTransformer(model_name_or_path, device=str(self.devices[0]))
+
+    def predict(self, query: str, documents: List[Document], top_k: Optional[int] = None) -> List[Document]:
+        """
+        Rank the documents based on their diversity and return the top_k documents.
+
+        :param query: The query.
+        :param documents: A list of Document objects that should be ranked.
+        :param top_k: The maximum number of documents to return.
+
+        :return: A list of top_k documents ranked based on diversity.
+        """
+        if query is None or len(query) == 0:
+            raise ValueError("Query is empty")
+        if documents is None or len(documents) == 0:
+            raise ValueError("No documents to choose from")
+
+        diversity_sorted = self.most_diverse_order(query=query, documents=documents)
+        return diversity_sorted[:top_k]
+
+    def most_diverse_order(self, query: str, documents: List[Document]) -> List[Document]:
+        """
+        Orders the given list of documents to maximize diversity. The algorithm first calculates embeddings for
+        each document and the query. It starts by selecting the document that is semantically closest to the query.
+        Then, for each remaining document, it selects the one that, on average, is least similar to the already
+        selected documents. This process continues until all documents are selected, resulting in a list where
+        each subsequent document contributes the most to the overall diversity of the selected set.
+
+        :param query: The search query.
+        :param documents: The list of Document objects to be ranked.
+
+        :return: A list of documents ordered to maximize diversity.
+        """
+
+        # Calculate embeddings
+        doc_embeddings: torch.Tensor = self.model.encode([d.content for d in documents], convert_to_tensor=True)
+        query_embedding: torch.Tensor = self.model.encode([query], convert_to_tensor=True)
+
+        n = len(documents)
+        selected: List[int] = []
+
+        # Compute the similarity matrix between documents
+        doc_sim_matrix: torch.Tensor = doc_embeddings @ doc_embeddings.T
+
+        # Compute the similarity vector between the query and documents
+        query_doc_sim: torch.Tensor = query_embedding @ doc_embeddings.T
```

Comment:
This only supports dot product similarity. I think using dot product similarity is recommended for the most part. However, I think other parts of haystack still support cosine similarity.